cli/internal/fs/package_deps_hash.go

```diff
@@ -31,354 +21,241 @@ type PackageDepsOptions struct {
 		} else {
-			filesToHash = append(filesToHash, filepath.Join(p.PackagePath, filename))
+			filesToHash = append(filesToHash, filePath.ToString())
 		}
 	}
-	normalized := make(map[string]string)
-	// These paths are platform-dependent, but already relative to the package root
-	for platformSpecificPath, hash := range result {
-		platformIndependentPath := filepath.ToSlash(platformSpecificPath)
-		normalized[platformIndependentPath] = hash
-	}
-	hashes, err := GetHashableDeps(filesToHash, p.PackagePath)
+
+	hashes, err := gitHashObject(rootPath, filesToHash)
 	if err != nil {
 		return nil, err
 	}
-	for platformIndependentPath, hash := range hashes {
-		normalized[platformIndependentPath] = hash
+
+	// Zip up file paths and hashes together
+	for filePath, hash := range hashes {
+		result[filePath] = hash
 	}
-	return normalized, nil
+
+	return result, nil
 }
 
 // GetHashableDeps hashes the list of given files, then returns a map of normalized path to hash
 // this map is suitable for cross-platform caching.
-func GetHashableDeps(absolutePaths []string, relativeTo string) (map[string]string, error) {
-	fileHashes, err := gitHashForFiles(absolutePaths)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to hash files %v", strings.Join(absolutePaths, ", "))
+func GetHashableDeps(rootPath AbsolutePath, files []string) (map[RelativeUnixPath]string, error) {
+	result, hashError := gitHashObject(rootPath, files)
+	if hashError != nil {
+		return nil, hashError
 	}
-	result := make(map[string]string)
-	for filename, hash := range fileHashes {
-		// Normalize path as POSIX-style and relative to "relativeTo"
-		relativePath, err := filepath.Rel(relativeTo, filename)
+
+	rootPathString := rootPath.ToString()
+	relativeResult := make(map[RelativeUnixPath]string)
+	for file, hash := range result {
+		relativePath, err := filepath.Rel(rootPathString, file.ToString())
 		if err != nil {
-			return nil, errors.Wrapf(err, "failed to get relative path from %v to %v", relativeTo, relativePath)
+			return nil, err
 		}
-		key := filepath.ToSlash(relativePath)
-		result[key] = hash
+		relativeResult[UnsafeToRelativeUnixPath(relativePath)] = hash
 	}
-	return result, nil
-}
 
-// threadsafeBufferWriter is a wrapper around a byte buffer and a lock, allowing
-// multiple goroutines to write to the same buffer. No attempt is made to
-// lock around reading, which should only be done once no more writing will occur.
-type threadsafeBufferWriter struct {
-	mu     sync.Mutex
-	buffer bytes.Buffer
+	return relativeResult, nil
 }
 
-func (tsbw *threadsafeBufferWriter) Write(p []byte) (int, error) {
-	tsbw.mu.Lock()
-	defer tsbw.mu.Unlock()
-	return tsbw.buffer.Write(p)
-}
+// gitHashObject takes a list of files returns a map of with their git hash values.
+// It uses git hash-object under the hood.
+// Note that filesToHash must have full paths.
+func gitHashObject(rootPath AbsolutePath, filesToHash []string) (map[RelativeUnixPath]string, error) {
```

Comment:
I need to clean up the comment. The paths that are passed in here should be `RelativeUnixPath`s and I should be able to push that type all the way to the call sites without too much pain. I'll do that.

Reply:
I've accidentally implemented a full path type system to make this happen and pushed it up to the call sites through most of the relevant code. We should talk about whether or not we should keep it, though, since it has runtime costs.